### PR TITLE
Fix depth scaling from in-place to out-of-place; add test cases in test_linear_algebra.jl

### PR DESCRIPTION
**Pull request by ziyiyin97**
Fix depth scaling from in-place to out-of-place; add test cases in test_linear_algebra.jl

### CODE REVIEW

**Comment by ziyiyin97** [comment]
This solves the issue in https://github.com/slimgroup/JUDI.jl/issues/40#issue-744176629

File: test/test_linear_algebra.jl [hunk 1]
```diff
@@ -94,4 +94,37 @@ fs =  parsed_args["fs"]
...
+
+        dobs1 = deepcopy(dobs)
+        dobs2 = deepcopy(dobs)
+        dobs3 = deepcopy(dobs)
+        dobs4 = deepcopy(dobs)
+
+        temp1 = F*dobs
+        temp2 = F'*dobs
+        temp3 = Md*dobs
+        temp4 = Md'*dobs
```
Comment (mloubout) [inline test/test_linear_algebra.jl:108]: that's not very efficient testing, it's only testing half of the issue. Also no idea what all these dobs2,3,4 are they don't serve any purpose.

```
dobs1 = deepcopy(dobs)
for Op in [F, F', Md , Md']
    m = Op*dobs 
     # Test that dobs wasn't modified
    @test dobs == dobs1
     # Test that it did compute something 
      @test m ~= dobs
```

same below for w

**Comment by mloubout** [comment]
Please add description to PR

File: test/test_linear_algebra.jl (hunk 1 above)
Comment (ziyiyin97) [reply test/test_linear_algebra.jl]: Modified both dobs and w in the latest commit

**Review by mloubout — COMMENTED** [review]
Looks fine, would be good to extend to 3D as well at some point.

**Comment by ziyiyin97** [comment]
Add 3D depth scaling operator and test associated.

File: src/TimeModeling/seismic_preconditioners.jl [hunk 2]
```diff
@@ -127,7 +127,7 @@ end
 
 function model_topmute(n::Tuple{Int64,Int64}, mute_end::Array{Integer,1}, length::Int64, x_orig)
 # Model domain topmute for a velocity model of dimensions n = [nx, nz].
-    x = copy(reshape(x_orig,n))
+    x = deepcopy(reshape(x_orig,n))
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:130]: 3D as well?

File: src/TimeModeling/seismic_preconditioners.jl [hunk 3]
```diff
@@ -191,11 +191,21 @@ end
 
 function depth_scaling(m, model)
 # Linear depth scaling function for seismic images
-    m = reshape(m,model.n)
+    m = deepcopy(reshape(m,model.n))
     filter = sqrt.(0f0:model.d[2]:model.d[2]*(model.n[2]-1))
     F = diagm(0=>filter)
-    for j=1:model.n[1]
-        m[j,:] = F*m[j,:]
+    if length(model.n) == 2
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:197]: There is probably a nicer way than this if else with joKron.

File: test/test_linear_algebra.jl [hunk 4]
```diff
@@ -94,4 +94,44 @@ fs =  parsed_args["fs"]
...
+        D3 = judiDepthScaling(model3D)
+
+        w2 = randn(Float32,prod(n))
+        w3 = deepcopy(w2)
+        for Op in [D3, D3']
+                m = Op*w2
+                # Test that w wasn't modified
+                @test w3 == w2
+                # Test that it did compute something 
+                @test m != w2
```
Comment (mloubout) [inline test/test_linear_algebra.jl:134]: test value of result as well

File: src/TimeModeling/seismic_preconditioners.jl [hunk 5]
```diff
@@ -191,11 +191,21 @@ end
 
 function depth_scaling(m, model)
 # Linear depth scaling function for seismic images
-    m = reshape(m,model.n)
+    m = deepcopy(reshape(m,model.n))
```
[the remark this answers is not in the document]
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl]: Fix this in the latest commit; add test which tests the value after applying depthscaling operator

File: src/TimeModeling/seismic_preconditioners.jl (hunk 2 above)
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl:130]: Done in the latest commit

File: src/TimeModeling/seismic_preconditioners.jl (hunk 3 above)
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl]: I've made a clean way to do depth scaling no matter 2D or 3D in the latest commit

File: test/test_linear_algebra.jl (hunk 4 above)
Comment (ziyiyin97) [reply test/test_linear_algebra.jl]: added value test of depth scaling and top mute

File: src/TimeModeling/seismic_preconditioners.jl [hunk 6]
```diff
@@ -189,15 +225,31 @@ function find_water_bottom(m)
     return idx
 end
 
+
+function find_water_bottom(m::Array{avDT,3};eps = 1e-4) where {avDT}
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:229]: Making your life lot harder. Always try to reuse as much as you have

`find_water_bottom(m::Array{avDT,3};eps = 1e-4) where {avDT} =  reshape(find_water_bottom(reshape(m, :, size(m, end));eps=eps), size(m,1), size(m,2))`

same blelow

File: src/TimeModeling/seismic_preconditioners.jl [hunk 7]
```diff
@@ -161,6 +159,45 @@ end
...
+    
+function model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Int64, length::Int64, x_orig)
+    # Model domain topmute for a velocity model of dimensions n = [nx, nz].
+    x = deepcopy(reshape(x_orig,n))
+    mute_start = mute_end - length
+    filter = zeros(Float32, n[end])
+    filter[1:mute_start-1] .= 0f0
+    filter[mute_end+1:end] .= 1f0
+    taper_length = mute_end - mute_start + 1
+    taper = (1f0 .+ sin.((Float32(pi)*(0:taper_length-1))/(taper_length - 1) .- Float32(pi)/2f0))/2f0
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:189]: similar to water bottom 

```
    mute_start = mute_end - length
    filter = zeros(Float32, n[end])
    filter[1:mute_start-1] .= 0f0
    filter[mute_end+1:end] .= 1f0
    taper_length = mute_end - mute_start + 1
    taper = (1f0 .+ sin.((Float32(pi)*(0:taper_length-1))/(taper_length - 1) .- Float32(pi)/2f0))/2f0
``` 

these lines are repeated four times, that's not a good sign.

**Review by mloubout — COMMENTED** [review]
Mostly their, thanks for your patience, few more comments, applies to the tests too.

File: src/TimeModeling/seismic_preconditioners.jl (hunk 6 above)
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl]: Yes clean up find_water_bottom for 3D in the latest commit; also added 2D/3D test for find_water_bottom function

File: src/TimeModeling/seismic_preconditioners.jl (hunk 7 above)
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl]: Now I implemented 3D top mute based on 2D top mute in the latest commit. Thank you for your advice!

File: src/TimeModeling/seismic_preconditioners.jl [hunk 8]
```diff
@@ -161,6 +161,29 @@ end
...
+
+function model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Array{Integer,2}, length::Int64, x_orig)
+    # Model domain topmute for a velocity model of dimensions n = [nx, nz].
+    x = deepcopy(reshape(x_orig,n))
+    x_out = deepcopy(reshape(x_orig,n))
+    for i = 1:n[2]
+        x_out[:,i,:] = reshape(model_topmute((n[1],n[3]),mute_end[:,i],length,x[:,i,:]),n[1],n[3])
+    end
+    return vec(x_out)
+end
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:173]: Still intricated and makes copy for each I

```
model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Array{Integer,2}, length::Int64, x_orig) = model_topmute((n[1]*n[2], n[3]), vec(mute_end), lengthx_orig)```

File: src/TimeModeling/seismic_preconditioners.jl [hunk 9]
```diff
@@ -161,6 +161,29 @@ end
...
+    
+function model_topmute(n::Tuple{Int64,Int64,Int64}, mute_end::Int64, length::Int64, x_orig)
+    # Model domain topmute for a velocity model of dimensions n = [nx, nz].
+    x = deepcopy(reshape(x_orig,n))
+    x_out = deepcopy(reshape(x_orig,n))
+    for i = 1:n[2]
+        x_out[:,i,:] = reshape(model_topmute((n[1],n[3]),mute_end,length,x[:,i,:]),n[1],n[3])
+    end
+    return vec(x_out)
+end
```
Comment (mloubout) [inline src/TimeModeling/seismic_preconditioners.jl:183]: same as above

File: src/TimeModeling/seismic_preconditioners.jl (hunk 8 above)
Comment (ziyiyin97) [reply src/TimeModeling/seismic_preconditioners.jl]: Thanks for your suggestion. Updated in the latest commit

**Pull request merged**